src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/TypesThatOwnDisposableFieldsShouldBeDisposable.cs

```diff
@@ -41,58 +41,64 @@ public override void Initialize(AnalysisContext analysisContext)
 
             analysisContext.RegisterCompilationStartAction(compilationContext =>
             {
-                if (!compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIDisposable, out var disposableType))
+                var wellKnownTypeProvider = WellKnownTypeProvider.GetOrCreate(compilationContext.Compilation);
+                if (!wellKnownTypeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIDisposable, out _))
                 {
                     return;
                 }
 
-                DisposableFieldAnalyzer analyzer = GetAnalyzer(disposableType);
+                DisposableFieldAnalyzer analyzer = GetAnalyzer(wellKnownTypeProvider);
                 compilationContext.RegisterSymbolAction(analyzer.AnalyzeSymbol, SymbolKind.NamedType);
             });
         }
 
-        protected abstract DisposableFieldAnalyzer GetAnalyzer(INamedTypeSymbol disposableType);
+        protected abstract DisposableFieldAnalyzer GetAnalyzer(WellKnownTypeProvider wellKnownTypeProvider);
 
         protected abstract class DisposableFieldAnalyzer
         {
             private readonly INamedTypeSymbol _disposableTypeSymbol;
+            private readonly INamedTypeSymbol _asyncDisposableTypeSymbol;
 
-            public DisposableFieldAnalyzer(INamedTypeSymbol disposableTypeSymbol)
+            public DisposableFieldAnalyzer(WellKnownTypeProvider wellKnownTypeProvider)
             {
-                _disposableTypeSymbol = disposableTypeSymbol;
+                _disposableTypeSymbol = wellKnownTypeProvider.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIDisposable);
+                _asyncDisposableTypeSymbol = wellKnownTypeProvider.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIAsyncDisposable);
             }
 
             public void AnalyzeSymbol(SymbolAnalysisContext symbolContext)
             {
                 INamedTypeSymbol namedType = (INamedTypeSymbol)symbolContext.Symbol;
-                if (!namedType.AllInterfaces.Contains(_disposableTypeSymbol))
+                if (namedType.AllInterfaces.Contains(_disposableTypeSymbol) ||
```

Comment:
I have added the second part of the `if` and reverted the condition to avoid deep nesting. The rest of the code of this method didn't changed except from the indentation.

Reply:
More efficient to write a single loop that walks through all interfaces and performs null check and then symbol equality check with both disposable type symbols. Note that `AllInterfaces` is an array, not a hashset.

Reply:
You are right I went for the easier to right not the optimal for the situation.